nutils/function.py

```diff
@@ -3276,6 +3276,136 @@ def _diagonalize(self, axis, newaxis):
+  shape : :class:`tuple` of :class:`int` or :class:`Array`
+      The shape of the array function.
+  dtype : :class:`bool`, :class:`int` or :class:`float`
+      The dtype of the array elements.
+
+  Example
+  -------
+
+  The following class implements ``multiply`` using :class:`Custom`
+  without broadcasting and for :class:`float` arrays only.
+
+  >>> class Multiply(Custom):
+  ...
+  ...   def __init__(self, left, right):
+  ...     left, right = broadcast_arrays(left, right)
+  ...     # Dtype coercion is beyond the scope of this example.
+  ...     if left.dtype != right.dtype:
+  ...       raise ValueError('left and right arguments should have the same dtype')
+  ...     super().__init__(args=(left, right), shape=left.shape, dtype=left.dtype)
+  ...
+  ...   @staticmethod
+  ...   def evalf(left, right):
+  ...     return left * right
+  ...
+  ...   @staticmethod
+  ...   def partial_derivative(iarg, left, right):
+  ...     if iarg == 0:
+  ...       return functools.reduce(diagonalize, range(left.ndim), right)
+  ...     elif iarg == 1:
+  ...       return functools.reduce(diagonalize, range(right.ndim), left)
+  ...     else:
+  ...       raise NotImplementedError
+  ...
+  >>> Multiply(asarray([1., 2.]), asarray([3., 4.])).eval()
+  array([[ 3.,  8.]])
+  >>> a = Argument('a', (2,))
+  >>> derivative(Multiply(a, asarray([3., 4.])), a).eval(a=numpy.array([1., 2.]))
+  array([[[ 3.,  0.],
+          [ 0.,  4.]]])
+
+  The following class implements the power of an array to a constant
+  :class:`int` scalar.
+
+  >>> class IntPower(Custom):
+  ...
+  ...   def __init__(self, base, exponent):
+  ...     # We are being nit-picky here and cast `exponent` to an `int` without
+  ...     # truncation.
+  ...     exponent = exponent.__index__()
+  ...     super().__init__(args=(base, exponent), shape=base.shape, dtype=float)
+  ...
+  ...   @staticmethod
+  ...   def evalf(base, exponent):
+  ...     # `base` is evaluated to a `numpy.ndarray` because we passed `base`
+  ...     # as an `Array` to the constructor. `exponent`, however, is untouched
+  ...     # because it is not an `Array`.
+  ...     return base**exponent
+  ...
+  ...   @staticmethod
+  ...   def partial_derivative(iarg, base, exponent):
+  ...     if iarg == 0:
+  ...       if exponent == 0:
+  ...         # The return value is cast to `Array` by the caller.
+  ...         return 0
+  ...       else:
+  ...         pd = exponent * IntPower(base, exponent - 1)
+  ...         return functools.reduce(diagonalize, range(base.ndim), pd)
+  ...     else:
+  ...       # We don't implement the derivative to `exponent`, because this is
+  ...       # a constant `int`.
+  ...       raise NotImplementedError
+  ...
+  >>> IntPower(asarray([2., 3.]), 2).eval()
+  array([[ 4.,  9.]])
+  >>> derivative(IntPower(a, 2), a).eval(a=numpy.array([2., 3.]))
+  array([[[ 4.,  0.],
+          [ 0.,  6.]]])
+  '''
+
+  def __init__(self, args, shape, dtype):
+    self._args = tuple(args)
```

Comment:
`_args` overrides the attribute of the `Immutable` base class, which affects simplify amongst others. I suggest changing it to `__args`.

Reply:
Tentatively added as 28a698f1.

Reply:
accepted